### PR TITLE
Feature/optional segment components

### PR DESCRIPTION
**Pull request by gnusi**
* Make segment reader components optional

### CODE REVIEW

File: core/index/segment_reader.hpp [hunk 1]
```diff
@@ -108,11 +105,9 @@ class segment_reader final : public sub_reader {
...
 
  private:
-  using impl_ptr = std::shared_ptr<const sub_reader>;
-
-  impl_ptr impl_;
+  std::shared_ptr<const sub_reader> impl_;
 
-  segment_reader(impl_ptr&& impl) noexcept;
-};  // segment_reade
+  segment_reader(std::shared_ptr<const sub_reader> impl) noexcept;
```
Comment (MBkkt) [inline core/index/segment_reader.hpp:110]: was rvalue only 
```suggestion
  segment_reader(std::shared_ptr<const sub_reader>&& impl) noexcept;
```

File: core/index/segment_reader.cpp [hunk 2]
```diff
@@ -243,7 +243,7 @@ segment_reader& segment_reader::operator=(
   const segment_reader& other) noexcept {
   if (this != &other) {
     // make a copy
-    impl_ptr impl = std::atomic_load(&other.impl_);
+    auto impl = std::atomic_load(&other.impl_);
```
Comment (MBkkt) [inline core/index/segment_reader.cpp:246]: NIT
load_explicit + memory_order

File: core/index/segment_reader.cpp [hunk 3]
```diff
@@ -259,7 +259,7 @@ segment_reader& segment_reader::operator=(
 
 segment_reader segment_reader::reopen(const segment_meta& meta) const {
   // make a copy
-  impl_ptr impl = std::atomic_load(&impl_);
+  auto impl = std::atomic_load(&impl_);
```
Comment (MBkkt) [inline core/index/segment_reader.cpp:262]: same here

**Review by Dronplane — APPROVED** [review]
LGTM

File: core/index/index_writer.hpp [hunk 4]
```diff
@@ -68,7 +68,7 @@ class readers_cache final : util::noncopyable {
 
   struct key_hash_t {
     size_t operator()(const key_t& key) const noexcept {
-      return hash_utils::Hash(key.name);
+      return absl::HashOf(key.name, key.version);
```
Comment (MBkkt) [inline core/index/index_writer.hpp:71]: revert?

**Review by MBkkt — APPROVED** [review]

**Pull request merged**